### ISSUE TITLE
SNI names must not end with a dot

### ISSUE THREAD

**Issue by ctron** [ctz/rustls#421]
In DNS it is valid to resolve `foo.bar.baz.`. Indicating that no search domains should be tried.

However, TLS SNI must not send this dot: https://tools.ietf.org/html/rfc6066#page-7

> "HostName" contains the fully qualified DNS hostname of the server, as understood by the client.  The hostname is represented as a byte string using ASCII encoding **without a trailing dot**.

`rusttls` sends the dot anyway:

~~~
ServerName(
    [
        ServerName {
            typ: HostName,
            payload: HostName(
                DNSName(
                    "keycloak.drogue-iot.svc.cluster.local.",   <---!!
                ),
            ),
        },
    ],
),
~~~

**Comment by Ralith** [comment]
Doesn't rustls just send whatever you tell it to?

**Comment by briansmith** [comment]
> Doesn't rustls just send whatever you tell it to?

I'm guessing it is fully relying on webpki's DNSName type to do the validation. Rustls might need to do some post-processing to remove the dot.

I filed https://github.com/briansmith/webpki/issues/144 about adding an API to webpki to help with this.

**Comment by ctz** [comment]
Thanks for the report; fixed in 20c4dab. I'll certainly move to using a webpki API for this later.
